tests/kabir_practice/matrix_multiplier.futil

```diff
@@ -0,0 +1,165 @@
+            lt2.left = col_index.out;
+            lt2.right = size.out;
+        }
+
+        comb group loop_guard_3 {   //Combinational guard for the inner tracker loop
+            lt3.left = loop_tracker.out;
+            lt3.right = size.out;
+        }
+
+        group reset_col {   //Reset column index to 0
+            col_index.in = 32'd0;
+            col_index.write_en = 1'd1;
+            reset_col[done] = col_index.done;
+        }
+
+        group reset_trackers {  //Reset loop tracker and cell total to 0
+            cell_total.in = 32'd0;
+            cell_total.write_en = 1'd1;
+
+            loop_tracker.in = cell_total.done ? 32'd0;
+            loop_tracker.write_en = cell_total.done ? 1'd1;
+
+            reset_trackers[done] = loop_tracker.done;
+        }
+        
+        group incr_col {    //Increment column index
+            adder.left = 32'd1;
+            adder.right = col_index.out;
+            col_index.in = adder.out;
+            col_index.write_en = 1'd1;
+            incr_col[done] = col_index.done;
+        }
+
+        group incr_row {    //Increment row index
+            adder.left = 32'd1;
+            adder.right = row_index.out;
+            row_index.in = adder.out;
+            row_index.write_en = 1'd1;
+            incr_row[done] = row_index.done;
+        }
+
+        group incr_loop_tracker {   //Increment loop tracker index
+            adder.left = 32'd1;
+            adder.right = loop_tracker.out;
+            loop_tracker.in = adder.out;
+            loop_tracker.write_en = 1'd1;
+            incr_loop_tracker[done] = loop_tracker.done;
+        }
+
+        group access_entries {  //From memory, access the latest pair of entries to multiply
+            //Position of left element
+            m1.addr0 = row_index.out;
+            m1.addr1 = loop_tracker.out;
+
+            //Position of right element
+            m2.addr0 = loop_tracker.out;
+            m2.addr1 = col_index.out;
+
+            m1_tracker.in = m1.read_data;
+            m1_tracker.write_en = 1'd1;
+
+            m2_tracker.in = m2.read_data;
+            m2_tracker.write_en = 1'd1;
+
+            access_entries[done] = m2_tracker.done;
+        }
+
+        group compute_cellwise_product { //Perform multiplication of two elements
+            multiplier.left = m1_tracker.out;
+            multiplier.right = m2_tracker.out;
+            multiplier.go = !multiplier.done ? 1'd1;
+
+            product_tracker.in = multiplier.done ? multiplier.out;
+            product_tracker.write_en = multiplier.done ? 1'd1;
+
+            compute_cellwise_product[done] = product_tracker.done;
+        }
+
+        group compute_entry { // Add latest cell product into accumulated total
+            adder.left = product_tracker.out;
+            adder.right = cell_total.out;
+
+            cell_total.in = adder.out;
+            cell_total.write_en = 1'd1;
+
+            compute_entry[done] = cell_total.done;
+        }
```

Comment:
Works great, but I just want to also make sure to teach an interesting feature: making a group do two things one after another. Your `compute_cellwise_product` will always be followed by `compute entry` because you are basically tying to do something like `foo = bar + baz * qux`. One option, over two groups, is:

1. `temp := baz * qux`
2. `foo := bar + temp`

And that's more or less what you have here.

The Calyx way of doing it as a one shot is this:

```
group mul_then_add {
  multiplier.left = baz.out
  multiplier.right = qux.out
  multiplier.go = 1
  
  adder.left = bar.out
  adder.right = multiplier.done ? multiplier.out // this needs to be conditional, since it won't make sense before the multiplier is done

  foo.in = multiplier.done ? adder.out // again needs to be conditional!
    // adder is combinational and gives an answer as soon as it has both its inputs.
    // But how do we know when adder has gotten its inputs? 
    // Well we can just remember: it gets its second input when multiplier.done is true.

  foo.write_en = multiplier.done ? 1
  
  mul_then_add[done] = foo.done
}
```

Reply:
CC @csziklai, since we talked about this only at a whiteboard.
CC @polybeandip, since you worked your way around this by using static stuff instead.